### PR TITLE
feat/subscription client

### PR DESCRIPTION
**Pull request by marco-ippolito**
I'm taking over @codeflyer PR https://github.com/mercurius-js/mercurius-subscription-client/pull/1
Linked to https://github.com/mercurius-js/mercurius/issues/929
resolves: https://github.com/mercurius-js/mercurius-subscription-client/issues/2

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,105 +1,371 @@
...
-
-## Install
-
-```bash
-npm i fastify mercurius graphql
-# or
-yarn add fastify mercurius graphql
+# @mercurius/subscription-client
+
+A `mercurius` plugin for `Subscriptions`.
```
Comment (codeflyer) [inline README.md:3]: This plugin manage just the client, the subscriptions are still managed inside mercurius.

File: lib/subscription-client.js [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+          if (this.keepAlive) {
+            this.startKeepAliveInterval()
+          }
+        } catch (err) {
+          this.close(this.tryReconnect, false)
+        }
+      }
+    }
+
+    this.socket.onclose = () => {
```
Comment (mcollina) [inline lib/subscription-client.js:80]: I think SubscriptionClient must inherit from EventEmitter and emit a `socketClose` event when this happen.

File: lib/subscription-client.js [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+      if (!this.closedByUser) {
+        this.close(this.tryReconnect, false)
+      }
+    }
+
+    this.socket.onerror = () => {}
```
Comment (mcollina) [inline lib/subscription-client.js:86]: I think SubscriptionClient must inherit from EventEmitter and emit a `socketError` event when this happens.

File: lib/subscription-client.js [hunk 4]
```diff
@@ -0,0 +1,338 @@
...
+    }
+    this.connect()
+  }
+
+  connect () {
+    this.socket = new WebSocket(this.uri, this.protocols)
+
+    this.socket.onopen = async () => {
+      /* istanbul ignore else */
+      if (this.socket && this.socket.readyState === WebSocket.OPEN) {
```
Comment (mcollina) [inline lib/subscription-client.js:65]: I think SubscriptionClient must inherit from EventEmitter and emit a `socketOpen` event when this happens.

File: lib/subscription-client.js [hunk 5]
```diff
@@ -0,0 +1,338 @@
...
+        }
+
+        /* istanbul ignore next */
+        throw new MER_ERR_GQL_SUBSCRIPTION_MESSAGE_INVALID(
+          `Invalid message type "${data.type}"`
+        )
+    }
+  }
+
+  startOperation (operationId) {
```
[the remark this answers is not in the document]
Comment (marco-ippolito) [reply lib/subscription-client.js]: I think this is supposed to be private

File: lib/subscription-client.js [hunk 6]
```diff
@@ -0,0 +1,338 @@
...
+    this.keepAlive = keepAlive
+
+    if (Array.isArray(protocols) && protocols.length > 0) {
+      this.protocols = protocols
+    } else {
+      this.protocols = [GRAPHQL_TRANSPORT_WS]
+    }
+
+    this.protocolMessageTypes = getProtocolByName(this.protocols[0])
+    this.keepAliveInterval = undefined
```
[the remark this answers is not in the document]
Comment (marco-ippolito) [reply lib/subscription-client.js]: I've made all internals `_`prefix, I've left user provided ones public

File: README.md [hunk 7]
```diff
@@ -1 +1,55 @@
-# mercurius-subscription-client
\ No newline at end of file
+# @mercurius/subscription-client
```
Comment (mcollina) [inline README.md:1]:
```suggestion
# @mercuriusjs/subscription-client
```

**Comment by marco-ippolito** [comment]
I was thinking abount implementing the event `ready` so it is possibile to `await once(client,ready)` to send the subscription. @mcollina  what do you think?

**Comment by simoneb** [comment]
@mcollina can I get some privileges on this repo and https://github.com/mercurius-js/mercurius-gateway please? Ideally the same I have on the main repo so we're somewhat independent in shipping these, triggering workflows etc

**Comment by mcollina** [comment]
You should be set up

**Review by mcollina — APPROVED** [review]
lgtm

File: test/subscription-client.js [hunk 8]
```diff
@@ -0,0 +1,841 @@
...
+  const operationId1 = client.createSubscription('query', {}, () => {})
+
+  client.unsubscribe(operationId1)
+  client.close()
+
+  await t.resolves(async () => await once(client, 'socketClose'))
+  server.close()
+})
+
+test('mercurius integration', async (t) => {
```
Comment (marco-ippolito) [inline test/subscription-client.js:725]: this is a integration test with mercurius. I'm not sure what should be the correct behaviour. For example `serviceName` if the mercurius is not part of a gateway ?
At the moment the test doesn't check anything because the createSubscription callback is never invoked and I can't figure why
 @mcollina  any idea?

**Comment by marco-ippolito** [comment]
I've tried to integrate this client with a mercurius server and for some reason the server never sends the response to the subscription request.
I've also tried with another client `graphql-ws` and had the same result.
The test is visible in the file `test/subscription-client.js` `mercurius integration`. I'm not sure what causes this behaviour

File: package.json [hunk 9]
```diff
@@ -0,0 +1,56 @@
+{
+  "name": "@mercuriusjs/subscription-client",
+  "version": "11.3.0",
```
Comment (mcollina) [inline package.json:3]:
```suggestion
  "version": "0.0.1",
```

File: test/subscription-client.js (hunk 8 above)
Comment (mcollina) [reply test/subscription-client.js:725]: What's the failure? I don't understand the problem.

Comment (marco-ippolito) [reply test/subscription-client.js:725]: this test: https://github.com/marco-ippolito/mercurius-subscription-client/blob/9080605208748b05a1d7a4172163e698496b0782/test/subscription-client.js#L725-L841
I'm trying to connect the client with a mercurius server.
When I create a subscription the `publish` callback is never called. I dont understand if I'm configuring something wrong or  there is some bug in the client.

File: package.json [hunk 10]
```diff
@@ -0,0 +1,56 @@
+{
+  "name": "@mercuriusjs/subscription-client",
+  "version": "0.0.1",
+  "description": "Fastify GraphQL adapter with gateway and subscription support",
```
Comment (jonnydgreen) [inline package.json:4]: should we make the description more tailored to this package?

File: test/types/index.ts [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+  serviceName: 'sample-name',
+  connectionCallback: async () => {
+  },
+  failedConnectionCallback: async () => {
+  },
+  failedReconnectCallback: async () => {
+  },
+  connectionInitPayload: {},
+  rewriteConnectionInitPayload: () => {
+  },
```
Comment (jonnydgreen) [inline test/types/index.ts:19]: Are there any parameters we can test the type of?

**Review by jonnydgreen — APPROVED** [review]
LGTM - just some non-blocking queries :)

File: SECURITY.md [hunk 12]
```diff
@@ -0,0 +1,69 @@
+# Security Policy
+
+This document describes the management of vulnerabilities for the
+Mercurius project and all modules within the Mercurius organization.
+
+## Reporting vulnerabilities
+
+Individuals who find potential vulnerabilities in Mercurius are invited
+to report them via GitHub at https://github.com/mercurius-js/mercurius/security/advisories
```
Comment (mcollina) [inline SECURITY.md:9]: This should point to this repo instead

File: _config.yml [hunk 13]
```diff
@@ -0,0 +1 @@
+theme: jekyll-theme-minimal
```
Comment (mcollina) [inline _config.yml:1]: is this file needed?

File: LICENSE [hunk 14]
```diff
@@ -0,0 +1,21 @@
+MIT License
+
+Copyright (c) 2018-2020 Matteo Collina and contributors
```
Comment (mcollina) [inline LICENSE:3]:
```suggestion
Copyright (c) Matteo Collina and contributors
```

File: package.json (hunk 10 above)
Comment (mcollina) [reply package.json]: @marco-ippolito this is not fixed

File: test/types/index.ts (hunk 11 above)
Comment (mcollina) [reply test/types/index.ts:19]: @marco-ippolito take a look

File: package.json (hunk 10 above)
Comment (marco-ippolito) [reply package.json]: I changed it to  ` "description": "GraphQL Subscription client",` maybe I should add more context?

Comment (mcollina) [reply package.json]: This is ok now

File: test/types/index.ts (hunk 11 above)
Comment (marco-ippolito) [reply test/types/index.ts:19]: @jonnydgreen  I added a few type tests
